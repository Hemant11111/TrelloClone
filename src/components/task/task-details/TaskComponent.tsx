import React, { useCallback, useState } from "react";
import { useDispatch } from "react-redux";

import "./Task.scss";
import TaskModel from "../../../model/Task";
import InputForm from "../../../shared-components/input-form/InputForm";
import { deleteTaskThunk, editTaskThunk } from "../../../redux/thunk/task.thunk";


interface TaskComponentProps {
    task: TaskModel;
}

export default function TaskComponent(props: TaskComponentProps) {

    const {task, task: {id, title, updating}} = props;

    const [_taskTitle, setTaskTitle] = useState(title);

    const dispatch = useDispatch();

    function _updateTask(task: TaskModel) {
        dispatch(editTaskThunk(task));
    }

    const handleDelete = useCallback(() => {
        if (window.confirm("Are you sure, you wants to delete this task?")) {
            dispatch(deleteTaskThunk(id));
        }
    }, []);

    const handleEdit = useCallback(() => {
        _updateTask({...task, updating: true});
    }, [task]);

    const handleUpdate = useCallback(() => {
        _updateTask({...task, title: _taskTitle, updating: false});
    }, [task]);

    const handleTaskChange = useCallback((e: any) => {
        setTaskTitle(e.target.value);
    }, []);


    const handleDragStart = useCallback((event: any) => {
        event.dataTransfer.setData("taskId", id);
    }, [])


    return (
        <div className="card task-container" draggable={true} onDragStart={handleDragStart}>
            {
                updating ? (
                    <InputForm
                        onChange={handleTaskChange}
                        onSubmit={handleUpdate}
                        value={_taskTitle}
                        placeholder="Add Task..."
                    />
                ) : _taskTitle
            }

            <div className="task-action-container">
                {updating ? (
                    <button
                        data-testid="task-save-button"
                        className="trello-icon-button"
                        onClick={handleUpdate}
                        disabled={!_taskTitle}
                    >
                        <i className="fa fa-save"/>
                    </button>
                ) : (
                    <button
                        data-testid="task-edit-button"
                        className="trello-icon-button" onClick={handleEdit}>
                        <i className="fa fa-edit"/>
                    </button>
                )}

                <button
                    data-testid="task-remove-button"
                    className="trello-icon-button task-delete" onClick={handleDelete}>
                    <i className="fa fa-remove"/>
                </button>
            </div>
        </div>
    )
}