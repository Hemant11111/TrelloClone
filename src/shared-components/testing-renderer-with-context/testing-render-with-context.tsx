import { ReactElement } from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";

import store from "../../redux/store/store";

export function testingRenderWithContext(element: ReactElement) {
    return render(
        <Provider store={store}>
            {element}
        </Provider>
    )
}