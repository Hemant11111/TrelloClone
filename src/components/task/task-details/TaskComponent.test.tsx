import React from 'react';
import { screen } from '@testing-library/react';

import TaskComponent from "./TaskComponent";
import TaskModel from "../../../model/Task";
import { testingRenderWithContext } from "../../../shared-components/testing-renderer-with-context/testing-render-with-context";


describe('Task Component', () => {

    it('should match snapshot', () => {
        const {asFragment} = testingRenderWithContext(<TaskComponent task={TaskModel.newInstance("test-task", "1")}/>);
        expect(asFragment()).toMatchSnapshot();
    })

    describe("when not editing", () => {

        const _task = TaskModel.newInstance("test-task", "1");

        it('should not render task update input when not editing a task', () => {
            testingRenderWithContext(<TaskComponent task={_task}/>);
            expect(screen.queryByPlaceholderText("Add Task...")).not.toBeInTheDocument();
        })

        it('should render remove task button when not editing', () => {
            testingRenderWithContext(<TaskComponent task={_task}/>);
            expect(screen.getByTestId("task-remove-button")).toBeInTheDocument();
        })

        it('should render edit task button', () => {
            testingRenderWithContext(<TaskComponent task={_task}/>);
            expect(screen.getByTestId("task-edit-button")).toBeInTheDocument();
        })

        it('should not render save task button', () => {
            testingRenderWithContext(<TaskComponent task={_task}/>);
            expect(screen.queryByTestId("task-save-button")).not.toBeInTheDocument();
        })
    })


    describe("when editing", () => {

        const _task = TaskModel.newInstance("test-task", "1");
        _task.updating = true;

        it('should render remove task button when editing', () => {
            testingRenderWithContext(<TaskComponent task={_task}/>);
            expect(screen.getByTestId("task-remove-button")).toBeInTheDocument();
        })

        it('should render save task button', () => {
            testingRenderWithContext(<TaskComponent task={_task}/>);
            expect(screen.getByTestId("task-save-button")).toBeInTheDocument();
        })

        it('should not render dit task button', () => {
            testingRenderWithContext(<TaskComponent task={_task}/>);
            expect(screen.queryByTestId("task-edit-button")).not.toBeInTheDocument();
        })

        it('should render task update input when editing a task', () => {
            testingRenderWithContext(<TaskComponent task={_task}/>);
            expect(screen.getByPlaceholderText("Add Task...")).toBeInTheDocument();
        })
    });
})