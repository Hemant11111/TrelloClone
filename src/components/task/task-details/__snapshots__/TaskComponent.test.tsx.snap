// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="card task-container"
    draggable="true"
  >
    test-task
    <div
      class="task-action-container"
    >
      <button
        class="trello-icon-button"
        data-testid="task-edit-button"
      >
        <i
          class="fa fa-edit"
        />
      </button>
      <button
        class="trello-icon-button task-delete"
        data-testid="task-remove-button"
      >
        <i
          class="fa fa-remove"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
